scripts/o3de/o3de/compatibility.py

```diff
@@ -0,0 +1,295 @@
+    gems_json_data = {}
+    for gem_path in gem_paths:
+        gem_path = pathlib.Path(gem_path).resolve()
+        gem_json_data = manifest.get_gem_json_data(gem_path=gem_path)
+        # only store the json data if it's a gem we're looking for and has compatibility data
+        if gem_json_data['gem_name'] in gem_names and\
+            (gem_json_data.get('compatible_engines') or gem_json_data.get('engine_api_dependencies')):
+            gems_json_data[gem_json_data['gem_name']] = gem_json_data
+
+    incompatible_gem_found = False
+    for gem_name in gem_names:
+        if gem_name not in gems_json_data:
+            continue
+        if not gem_engine_compatible(gems_json_data[gem_name], engine_json_data, gem_paths):
+            incompatible_gem_found = True
+            logger.error(f'{gem_name} is not known to be compatible with the '
+                            f'engine {engine_json_data["engine_name"]} and requires the --force '
+                            'parameter to register.')
+
+    return not incompatible_gem_found
+
+
+def gem_dependencies_compatible(gem_json_data:dict, project_path:pathlib.Path = None, gem_paths:list = [], check:bool = False) -> bool:
+    """
+    Returns True if the gem is compatible with the gems registered with the engine and optional project.
+    :param gem_json_data: gem json data dictionary
+    :param project_path: path to the project (optional)
+    :param gem_paths: paths to all gems to use for dependency checks (optional)
+    :param check: if True always return True and ouput log info about the result
+    """
+    # try to avoid gem compatibility checks which incur the cost of 
+    # opening many gem.json files to get version information
+    gem_dependencies = gem_json_data.get('dependencies','')
+    if not gem_dependencies:
+        return True
+
+    incompatible_gem_version_specifiers = None
+
+    if not gem_paths:
+        gem_paths = manifest.get_engine_gems()
+        gem_paths.extend(manifest.get_project_gems(project_path))
+        # Convert each path to pathlib.Path object and filter out duplicates using dict.fromkeys
+        gem_paths = list(dict.fromkeys(map(lambda gem_path_string: pathlib.Path(gem_path_string), gem_paths)))
+
+    incompatible_gem_version_specifiers = get_incompatible_gem_version_specifiers(
+        project_path, gem_dependencies, gem_paths)
+    if incompatible_gem_version_specifiers:
+        if check:
+            logger.info(f'{gem_json_data["gem_name"]} is not known to be compatible with the '
+            ' following gems used by this project and requires the --force parameter to enable.'
+            f'{incompatible_gem_version_specifiers}')
+        else:
+            logger.error(f'{gem_json_data["gem_name"]} is not known to be compatible with the '
+            ' following gems used by this project and requires the --force parameter to enable.'
+            f'{incompatible_gem_version_specifiers}')
```

Comment:
So it seems like this message should be part of enable gem, since that that is where the activation code is located at.